go-controller/pkg/metrics/service.go

```diff
@@ -0,0 +1,93 @@
+package metrics
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+var metricServiceCount = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+	Namespace: MetricOvnNamespace,
+	Subsystem: MetricOvnSubsystemController,
+	Name:      "service_count",
+	Help: "Metric that specifies the number of services by their type. " +
+		"Type of the service is specified with the label type.",
+},
+	[]string{
+		"type",
+	},
+)
+
+type ServiceMetrics interface {
+	Increment(service *kapi.Service)
+	Decrement(service *kapi.Service)
+	Get(service *kapi.Service) uint64
+	GetAll() map[kapi.ServiceType]uint64
+}
+
+type serviceMetricsData struct {
+	sync.RWMutex
+	metricsMap map[kapi.ServiceType]uint64
+}
+
+var serviceMetrics ServiceMetrics = &serviceMetricsData{metricsMap: make(map[kapi.ServiceType]uint64)}
+
+func GetServiceMetricsInstance() ServiceMetrics {
+	return serviceMetrics
+}
+
+func (s *serviceMetricsData) Increment(service *kapi.Service) {
+	s.Lock()
+	defer s.Unlock()
+	if s.metricsMap[service.Spec.Type] < ^uint64(0) {
+		s.metricsMap[service.Spec.Type]++
+	}
+}
+
+func (s *serviceMetricsData) Decrement(service *kapi.Service) {
+	s.Lock()
+	defer s.Unlock()
+	if s.metricsMap[service.Spec.Type] > 0 {
+		s.metricsMap[service.Spec.Type]--
+	}
+}
+
+func (s *serviceMetricsData) Get(service *kapi.Service) uint64 {
+	s.RLock()
+	defer s.RUnlock()
+	return s.metricsMap[service.Spec.Type]
+}
+
+func (s *serviceMetricsData) GetAll() map[kapi.ServiceType]uint64 {
```

Comment:
I have done it this way with the following in mind:
1. Map is a reference type in GOLANG just like pointers.  This is the fundamental reason why I took this approach.
2. You do not want to return a reference to a data structure from the API which is mutable. General API design should make sure data is protected without making any assumptions. 
3. Point no 2 is especially true in case of concurrent scenarios in which case passing a reference provides the unprotected access to the data structure which is true in our case. This will result in race conditions.